cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go

```diff
@@ -21,38 +21,130 @@ import (
 	"regexp"
 	"strconv"
 	"strings"
+	"sync"
+	"time"
 
+	aws_util "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
 	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
 	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	kube_client "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
 )
 
 // AwsCloudProvider implements CloudProvider interface.
 type AwsCloudProvider struct {
 	awsManager *AwsManager
+	asgMutex   sync.Mutex
 	asgs       []*Asg
 }
 
 // BuildAwsCloudProvider builds CloudProvider implementation for AWS.
-func BuildAwsCloudProvider(awsManager *AwsManager, specs []string) (*AwsCloudProvider, error) {
+func BuildAwsCloudProvider(awsManager *AwsManager, specs []string, kubeClient kube_client.Interface, autoDiscoverNodeGroup bool) (*AwsCloudProvider, error) {
 	aws := &AwsCloudProvider{
 		awsManager: awsManager,
 		asgs:       make([]*Asg, 0),
 	}
-	for _, spec := range specs {
-		if err := aws.addNodeGroup(spec); err != nil {
-			return nil, err
+	if autoDiscoverNodeGroup {
+		go func() {
+			for {
+				selector := "master!=true"
+				listAll := apiv1.ListOptions{LabelSelector: selector}
+				nl, err := kubeClient.Core().Nodes().List(listAll)
+				if err != nil {
+					glog.Errorf("err: %s\n", err)
+				}
+
+				var nodesList []*string
+
+				for _, n := range nl.Items {
+					glog.V(4).Infof("Considering node: %s", n.Name)
+					nodesList = append(nodesList, &n.Name)
+				}
+				AutoDiscoverNodeGroup(aws, nodesList)
+				time.Sleep(30 * time.Second)
```

Comment:
Could it run in the same loop [here](https://github.com/kubernetes/contrib/blob/master/cluster-autoscaler/cluster_autoscaler.go#L227)